src/librustc_typeck/collect.rs

```diff
@@ -447,6 +449,52 @@ impl AstConv<'tcx> for ItemCtxt<'tcx> {
     }
 }
 
+/// Synthesize a new lifetime name that doesn't clash with any of the lifetimes already present.
+/// It will pick a single `char` lifetime name that is available, or some `lt_{number}` that is
+/// not already in use.
+fn get_new_lifetime_name<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    poly_trait_ref: ty::PolyTraitRef<'tcx>,
+    generics: &hir::Generics<'tcx>,
+) -> String {
+    let mut lt_char = 'a';
+    let mut lt_name = format!("'{}", lt_char);
+    let mut existing_lifetimes = tcx
+        .collect_referenced_late_bound_regions(&poly_trait_ref)
+        .into_iter()
+        .filter_map(|lt| {
+            if let ty::BoundRegion::BrNamed(_, name) = lt {
+                Some(name.as_str().to_string())
+            } else {
+                None
+            }
+        })
+        .chain(generics.params.iter().filter_map(|param| {
+            if let hir::GenericParamKind::Lifetime { .. } = &param.kind {
+                Some(param.name.ident().as_str().to_string())
+            } else {
+                None
+            }
+        }))
+        .collect::<Vec<String>>();
+    existing_lifetimes.sort();
+    while existing_lifetimes.contains(&lt_name) {
+        let prev = lt_char;
+        lt_char = std::char::from_u32(lt_char as u32 + 1).unwrap_or(lt_char);
```

Comment:
What about using [repeated characters](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8c25cc9587846b4dc5a4230adda8602b) instead of numbers? Something like:

```rust
let a_to_z_repeat_n = |n| {
    (b'a'..=b'z').map(move |c| {
        let mut ret = format!("'");
        ret.extend(iter::repeat(char::from(c)).take(n));
        ret
    })
};

let lt_name = (1..)
    .flat_map(a_to_z_repeat_n)
    .find(|lt| !existing_lifetimes.contains(lt))
    .unwrap();
```

Reply:
done